### ISSUE TITLE
Jack2 on Windows 10 crashes when stopped

### ISSUE THREAD

**Issue by ReinholdH** [jackaudio/jack2#648]
With Jack2 1.9.16 x64 (jack2-win64-v1.9.16.exe)

When stopping Jack2 a crash message appreas in the message box. See message attached.
[Crash_message_when_Jack_is_stopped.txt](https://github.com/jackaudio/jack2/files/5395175/Crash_message_when_Jack_is_stopped.txt)

**Comment by kmatheussen** [comment]
There was a problem in earlier versions of QJackCtl that it always reported
that jack had "crashed" after jack had stopped running (jack didn't
actually crash). That problem should have been fixed though, but maybe not.
Does it happen if you use the dummy driver?

**Comment by ReinholdH** [comment]
It happens with the default setting.

**Comment by kmatheussen** [comment]
Can you try the dummy driver? It's to check if it's the portaudio
driver that crashes. I don't have a windows computer in front of me
now so I can't check myself.

**Comment by kmatheussen** [comment]
(I meant portaudio interface, not portaudio driver. portaudio itself
has probably not crashed, but the interface can be more sketchy,
depending on your soundcard driver and so forth.)

**Comment by ReinholdH** [comment]
Thx, but not familiar how to setup a dummy interface on Windows that is visible in QjackCtl. Any help is appreciated.

**Comment by kmatheussen** [comment]
Hi Reinhold,

You select the "dummy" driver under "Setup". But I've tried now, and
qjackctl reports that it crashes even when using the dummy driver too. I'm
pretty certain this is a bug in qjackctl, one you can safely ignore, and
not a bug in jack. I fixed this bug a long time ago, but it seems to have
reappeared.

**Comment by kmatheussen** [comment]
Can you try this older version of qjackctl and see if it reports that jack
is crashing? (I'm not able to make it run under wine, but it should work
under windows)

http://users.notam02.no/~kjetism/qjackctl.exe

**Comment by kmatheussen** [comment]
Also, can you check the version of qjackctl you are running? Perhaps
you are running an older version of qjackctl for some reason, and that
we always see the crash message under wine for some reason (causing me
to see the same error message).

@rncbc perhaps we should never display the "Jack crashed" message in
qjackctl after jack has exited anyway? I think it's usually only
caused by QProcess misreporting that something went wrong.

**Comment by ReinholdH** [comment]
Thx. I tried your older version of qjackctl and it does not crash when stopped. The old one is fine.
I use the official qjackctl which is included in jack2-win64-v1.9.16.exe. The qjackctl version is 0.6.3 using JACK 1.9.16.

When stopping Jack the message is displayed
JACK was stopped
Jack has crashed

**Comment by kmatheussen** [comment]
Thank you, I'll investigate what happened. Perhaps that patch was never
merged upstream. Also, just want to point out again that nothing crashes
here, it's just the message that is wrong, most likely.

**Comment by kmatheussen** [comment]
@rnbc The offending commit seems to be this one:
66e63639d71b6ba56e03e734e003cb85c56a78a5

More specifically it undid this patch:

+ #if defined(WIN32)
+ if (!m_bJackStopped)
+#endif
          appendMessagesColor(tr("JACK has crashed."), "#cc3366");

**Comment by kmatheussen** [comment]
Well, I guess it's possible that jack crashes, but the user shouldn't
have to know about this. It doesn't matter that it crashes at exit. It
just looks unprofessional. In my opinion, the error message should be
removed first, and then we might investigate why qprocess reports that
it crashed.

**Comment by kmatheussen** [comment]
No, this is a qjackctl thing only. qjackctl kill()s jackd on windows,
and that's probably why qprocess reports that it crashes:

#if defined(__WIN32__) || defined(_WIN32) || defined(WIN32)
// Try harder...
m_pJack->kill();
#else
// Try softly...
m_pJack->terminate();
#endif

So this is probably a good enough solution:

+ #if defined(WIN32)
+ if (!m_bJackStopped)
+#endif
          appendMessagesColor(tr("JACK has crashed."), "#cc3366");

**Comment by kmatheussen** [comment]
And replace WIN32 with _WIN32:

 +#if defined(_WIN32)
+   if (!m_bJackStopped)
+#endif
          appendMessagesColor(tr("JACK has crashed."), "#cc3366");

**Comment by falkTX** [comment]
I removed that specific win32 bit because the logic that qjackctl had for a variable that is set as true when user clicks to stop jack is no longer there.
The other, new variables related to something else (jack client shutdown), which is not the same and would lead to confusion.

Cant we just make qjackctl call a proper terminate on the process first, before attempting to kill it?
Making jackd.exe not having a chance to clean itself up is pretty bad IMO, this is why we have issues with "jack_db" as it is never removed (a step done at jackd exit)

**Comment by kmatheussen** [comment]
I guess kill() is called because jackd doesn't listen to WM_CLOSE messages.
From the QProcesss:terminal documentation:

"On Windows, terminate() posts a WM_CLOSE message to all top-level windows
of the process and then to the main thread of the process itself. On Unix
and macOS the SIGTERM signal is sent.

Console applications on Windows that do not run an event loop, or whose
event loop does not handle the WM_CLOSE message, can only be terminated by
calling kill <https://doc.qt.io/qt-5/qprocess.html#kill>().
"

**Comment by falkTX** [comment]
Last time I checked, jackd listens to the ctrl_c CLI thing, which is similar to WM_CLOSE.
But need to retry this

**Comment by falkTX** [comment]
The relevant code: https://github.com/jackaudio/jack2/blob/develop/common/JackControlAPI.cpp#L537

**Comment by kmatheussen** [comment]
Could be. I haven't tried. I guess @sletz is the one writing this code
in qjackctl. Maybe he remembers why that code is there?

**Comment by sletz** [comment]
I guess this code is there because it is compiled in libjackserver DLL which is then used by the jackdmp binary.

**Comment by kmatheussen** [comment]
The commit that added
https://github.com/jackaudio/jack2/blob/develop/common/JackControlAPI.cpp#L537
(e44ab5b9ade5bfaa49f1a9d3f7fa8f08893054e9) was made in 2012.  The kall to
kill() in qjackctl was probably implemented much earlier than 2012, so I
guess it's very likely that the call to kill() on windows can be removed
now (
https://github.com/rncbc/qjackctl/blob/master/src/qjackctlMainForm.cpp#L1641
).

**Comment by kmatheussen** [comment]
Sorry, a6a2d8f0e6a8b0df82527ea71183fe6bf45d6933, was just a refactoring. I
can't find the original commit, but it's from 2008 or older.

**Comment by rncbc** [comment]
just raise a PR to https://github.com/rncbc/qjackctl when you conclude your investigations :)
i'll be glad to blindly merge whatever (as long it only affects the windows build).
cheers

**Comment by rncbc** [comment]
anyway, https://github.com/rncbc/qjackctl/commit/0aadfd4 is probably a comeback to `m_bJackStopped` variable, neé  `m_bJackKilled`... ;)

fair warning: not tested whatsoever in the field: counting on you Windows freaks to do just that

cheers

**Comment by falkTX** [comment]
Would be best to try to stop jack nicely first. But that is still a nice commit to have in case the clean exit fails and qjackctl ends up calling "kill" anyway.
I will do some tests regarding trying a clean stop soon.

**Comment by rncbc** [comment]
as far as I can remember, and probably @sletz maybe also recall, invoking QProcess:terminate() as we/I do in good old Linux is just a no-op in windows (yeah, it sends and a dang WM_CLOSE message to the target process and, omg, there's no event-loop whatsoever!

bummer

**Comment by sletz** [comment]
No precise remembering on any specific QT thing, I'm afraid.

**Comment by rncbc** [comment]
@sletz, np. 

anyway _it is_ all right there documented in the Qt docs: 
https://doc.qt.io/qt-5/qprocess.html#terminate

**Comment by falkTX** [comment]
That is odd, according to official docs there are other ways to handle this.
https://docs.microsoft.com/en-us/windows/win32/procthread/terminating-a-process

In any case, we just need to find something that works.
Creating a dummy handle just to receive WM_CLOSE is similar to what lbjackserver is doing already, so we can just expand that.

**Comment by rncbc** [comment]
i wouldn't go anyway far just for silencing a dang and annoying line on the qjackctl messages... but that's your call anyway :)

**Comment by falkTX** [comment]
It is not just about silencing a "stop", it is killing jackd instead of letting it nicely close.
This leads to issues when trying to start jackd again.

**Comment by rncbc** [comment]
have you tested it? does the "JACK has crashed" message _ever_ resulted in a no restartable jackd?

**Comment by falkTX** [comment]
Yes, killing jackd can lead to not being able to start it again (until a reboot)
Somewhat relevant #647
It is during cleanup/stop that jackd deletes the meta-data DB stuff.

Anyway, it is just a good practice to stop processes cleanly.
